### PR TITLE
chore: add Postgres 15 to CI

### PR DESCRIPTION
**Pull request by jmealo**
## Description
To avoid compilation issues with `wal2json`, we can use the `pgdg` package to install it instead.

## Performance impact
- None

## Security impact
- None

## Checklist

- [x] My code matches the project's code style and `yarn lint:fix` passes.
- [x] I've added tests for the new feature, and `yarn test` passes.
- [x] I have detailed the new feature in the relevant documentation.
- [x] I have added this feature to 'Pending' in the `RELEASE_NOTES.md` file (if one exists).
- [x] If this is a breaking change I've explained why.

### CODE REVIEW

**Comment by jmealo** [comment]
@benjie: Hopefully this fixes it! I'll need a workflow approval.

**Comment by jmealo** [comment]
@benjie: I would consider using the `pgdg`debian/apt repos to install logical decoding plugins since you're using a Debian base image anyway. That way you can stick to a given version instead of baking in the commit hash. They offer `arm64` and `amd64` binaries so it's pretty awesome :) 


Then you can just use: `apt-get install postgresql-${PG_VERSION}-wal2json`

**Comment by jmealo** [comment]
Ok, on my fork, prior to making the change to use the `pgdg` package for `wal2json`, the CI fails on 15 because of this:
```
Run sudo bash -c "echo deb http://apt.postgresql.org/pub/repos/apt/ focal-pgdg main >> /etc/apt/sources.list.d/pgdg.list"
  sudo bash -c "echo deb http://apt.postgresql.org/pub/repos/apt/ focal-pgdg main >> /etc/apt/sources.list.d/pgdg.list"
  wget --quiet -O - https://www.postgresql.org/media/keys/ACCC4CF8.asc | sudo apt-key add -
  sudo apt-get update
  sudo apt-get -yqq install postgresql-client-15
  shell: /usr/bin/bash -e {0}
  env:
    CI: true
    PGVERSION: 15
    TEST_DATABASE_URL: ***localhost:543[2](https://github.com/jmealo/postgraphile/actions/runs/5061798040/jobs/9086496801#step:6:2)/graphileengine_test
    TEST_PG_URL: ***localhost:5432/postgraphile_test
    LDS_TEST_DATABASE_URL: ***localhost:5432/lds_test
    TERM: xterm
    FORCE_COLOR: 1
Warning: apt-key is deprecated. Manage keyring files in trusted.gpg.d instead (see apt-key(8)).
OK
Hit:1 http://azure.archive.ubuntu.com/ubuntu jammy InRelease
Get:2 http://azure.archive.ubuntu.com/ubuntu jammy-updates InRelease [119 kB]
Get:3 http://azure.archive.ubuntu.com/ubuntu jammy-backports InRelease [108 kB]
Get:4 http://azure.archive.ubuntu.com/ubuntu jammy-security InRelease [110 kB]
Get:5 https://packages.microsoft.com/ubuntu/22.04/prod jammy InRelease [3611 B]
Get:6 http://azure.archive.ubuntu.com/ubuntu jammy-updates/main amd64 Packages [609 kB]
Get:7 http://azure.archive.ubuntu.com/ubuntu jammy-updates/main Translation-en [172 kB]
Get:8 http://azure.archive.ubuntu.com/ubuntu jammy-updates/main amd64 c-n-f Metadata [14.6 kB]
Get:9 http://azure.archive.ubuntu.com/ubuntu jammy-updates/restricted amd64 Packages [260 kB]
Get:10 http://azure.archive.ubuntu.com/ubuntu jammy-updates/restricted Translation-en [38.7 kB]
Get:11 http://azure.archive.ubuntu.com/ubuntu jammy-updates/universe amd64 Packages [906 kB]
Get:12 http://azure.archive.ubuntu.com/ubuntu jammy-updates/universe Translation-en [186 kB]
Get:13 http://azure.archive.ubuntu.com/ubuntu jammy-updates/universe amd64 c-n-f Metadata [18.9 kB]
Get:14 http://azure.archive.ubuntu.com/ubuntu jammy-updates/multiverse amd64 Packages [35.3 kB]
Get:15 http://azure.archive.ubuntu.com/ubuntu jammy-updates/multiverse Translation-en [8452 B]
Get:16 http://apt.postgresql.org/pub/repos/apt focal-pgdg InRelease [117 kB]
Get:17 http://azure.archive.ubuntu.com/ubuntu jammy-backports/universe amd64 Packages [23.3 kB]
Get:18 http://azure.archive.ubuntu.com/ubuntu jammy-backports/universe amd64 c-n-f Metadata [548 B]
Get:19 http://azure.archive.ubuntu.com/ubuntu jammy-security/main amd64 Packages [394 kB]
Get:20 http://azure.archive.ubuntu.com/ubuntu jammy-security/main Translation-en [113 kB]
Get:21 http://azure.archive.ubuntu.com/ubuntu jammy-security/main amd64 c-n-f Metadata [9884 B]
Get:22 http://azure.archive.ubuntu.com/ubuntu jammy-security/restricted amd64 Packages [260 kB]
Get:23 http://azure.archive.ubuntu.com/ubuntu jammy-security/restricted Translation-en [38.3 kB]
Get:24 http://azure.archive.ubuntu.com/ubuntu jammy-security/universe amd64 Packages [727 kB]
Get:25 http://azure.archive.ubuntu.com/ubuntu jammy-security/universe Translation-en [127 kB]
Get:26 http://azure.archive.ubuntu.com/ubuntu jammy-security/universe amd64 c-n-f Metadata [14.6 kB]
Get:27 http://azure.archive.ubuntu.com/ubuntu jammy-security/multiverse amd64 Packages [30.2 kB]
Get:28 http://azure.archive.ubuntu.com/ubuntu jammy-security/multiverse Translation-en [5828 B]
Get:29 https://ppa.launchpadcontent.net/ubuntu-toolchain-r/test/ubuntu jammy InRelease [23.8 kB]
Get:30 https://packages.microsoft.com/ubuntu/22.04/prod jammy/main armhf Packages [7357 B]
Get:31 https://packages.microsoft.com/ubuntu/22.04/prod jammy/main amd64 Packages [65.0 kB]
Get:32 https://packages.microsoft.com/ubuntu/22.04/prod jammy/main all Packages [904 B]
Get:33 https://packages.microsoft.com/ubuntu/22.04/prod jammy/main arm64 Packages [14.2 kB]
Get:34 http://apt.postgresql.org/pub/repos/apt focal-pgdg/main amd64 Packages [264 kB]
Get:35 https://ppa.launchpadcontent.net/ubuntu-toolchain-r/test/ubuntu jammy/main amd64 Packages [15.7 kB]
Get:36 https://ppa.launchpadcontent.net/ubuntu-toolchain-r/test/ubuntu jammy/main Translation-en [7292 B]
Fetched 4845 kB in 2s (1996 kB/s)
Reading package lists...
W: http://apt.postgresql.org/pub/repos/apt/dists/focal-pgdg/InRelease: Key is stored in legacy trusted.gpg keyring (/etc/apt/trusted.gpg), see the DEPRECATION section in apt-key([8](https://github.com/jmealo/postgraphile/actions/runs/5061798040/jobs/9086496801#step:6:8)) for details.
E: Unable to correct problems, you have held broken packages.
Error: Process completed with exit code [10](https://github.com/jmealo/postgraphile/actions/runs/5061798040/jobs/9086496801#step:6:10)0.
```

**Comment by jmealo** [comment]
@benjie Hey! I cleaned up the CI a little bit and updated dependencies where needed to support newer versions of Node.  I'm not sure if you want to keep running tests for deprecated versions of Node, but, AFAICT, the tests should still pass.

**Comment by jmealo** [comment]
I'm about to land so I'll be AFK.

- My PR fixes PostgreSQL 15. We need to pass `NODE_OPTIONS` with a legacy crypto flag to fix webpack. We must do it conditionally, 17.x and up or node will fail to start because of the unrecognized flag.

- We could consider running build on 16.x and tests on each version if it ends up being an issue.

- Upgrading webpack isn't a simple version bump, the current configuration doesn't pass schema validation.

**Comment by jmealo** [comment]
Ooops. This didn't send yesterday:

- My PR fixes PostgreSQL 15. I should probably do one major dependency at a time but here we are.


- We could consider running build on 16.x and tests on each version if it ends up being an issue.

- Upgrading webpack isn't a simple version bump, the current configuration doesn't pass schema validation.

Why this is painful:
We need to pass `NODE_OPTIONS` with a legacy crypto flag to fix webpack. We must do it conditionally@gt 17.x and up OR ELSE NODE DIES ON START-UP  because of the unrecognized flag. There's so many different node invocations and different GitHub actions use different versions of node so getting the correct combination of env required to avoid upgrading webpack makes me want to extract the issue entirely.

What are your thoughts on some possible approaches for either isolating/extracting webpack to where it wants to run vs trying to fix it in place?


- Would you rather attempt to use a newer version of create-react-app and swap in your existing code and hope it works or troubleshoot upgrading the webpack config?

- Use a dedicated build step/job that's running Node 16 and pass the build/dist output as an input to the various test environments. This seems like the easiest and we're not testing webpack here so this would make the build more resilient. ✨

Do folks need to build in their environment or could this happen only in CI? I'm not familiar - with the React portion and it caught me by surprise. 🤦‍♂️

File: scripts/test
Comment (benjie) [inline scripts/test:2]: Typo?

Comment (jmealo) [reply scripts/test:2]: Yes! Sorry, was on a plane preparing to land 🤦  I'll take a look at this again tonight.

**Comment by jmealo** [comment]
@benjie: `/scripts/build` doesn't run for me locally. I'm having issues separating the build/test commands because I'm apparently not capturing all of the output? I think it's the favico? It seems like we're pushing stuff back into `/src`? Also, getting tons of Webpack errors (even without upgrading) due to the 1MiB html file. 

What confuses the heck out of me, is even when I run Node 16.20.0 and blow out my `node_modules`, I get 2 typescript errors :shrug:

**Comment by jmealo** [comment]
@benjie: I'm not sure if you can see the CI runs on my fork, but I have this working finally... Unfortunately, it looks like there is an actual issue running on Node 18/20.

**Comment by jmealo** [comment]
Ah, I removed Node 18, let's see how that goes 🤞  ...

**Comment by jmealo** [comment]
@benjie: We're all green here, this should be good to merge, when you decide to opt-in for Node 18,20 ... you can continue to build on 16, and avoid the webpack upgrade.

File: .github/workflows/ci.yml
Comment (benjie) [inline .github/workflows/ci.yml:28]: I don't see the formatting and linting rules in the new CI file; am I missing something?

File: .github/workflows/ci.yml
Comment (benjie) [inline .github/workflows/ci.yml:33]: I've found previously that uploading and then downloading artifacts like this can be slower than just rebuilding everything in the first place because GitHub Action's caches are so slow. Since only one thing depends on this (`test`), I'm not sure why it's necessary? Perhaps this is because you planned to add the `lint` task as well?

**Review by benjie — CHANGES_REQUESTED** [review]
Sorry for the delay, I took some time off and then caught covid (first time).

I think linting got dropped?

File: .github/workflows/ci.yml
Comment (jmealo) [reply .github/workflows/ci.yml:33]: @benjie Hey! I hope you're feeling better. This change was made because the current webpack is hard to upgrade and won't run on versions of Node greater than 16. By locking it in to run on the latest version that it's compatible with unlocks the potential to run 18/20 in CI without replacing Webpack (which I did attempt and it was very messy, for me at least).

**Comment by jmealo** [comment]
> > Use a dedicated build step/job that's running Node 16 and pass the build/dist output as an input to the various test environments. This seems like the easiest and we're not testing webpack here so this would make the build more resilient. ✨
> 
> This would be preferred. PostGraphile V4 is essentially in maintenance mode as we ramp up the final stages of PostGraphile V5 development, so I really would rather avoid changing anything significant about V4 unless it actually stops working.


^^^ So this is why I broke out build into a separate job to feed the test step. I removed Node 18/20 from this PR, but, the groundwork has been done for supporting them.

File: .github/workflows/ci.yml
Comment (jmealo) [reply .github/workflows/ci.yml]: Ooops. You're right, I added them to the build step.

**Comment by jmealo** [comment]
I added back in the lint step (ooops). I think this should be good.

**Comment by jmealo** [comment]
Let me know if you have any other questions (answered above) and whether you'd like me to squash my commits.

File: .github/workflows/ci.yml
Comment (benjie) [reply .github/workflows/ci.yml:37]: Ah yes, I remember you saying about this before. Sorry for forgetting!

**Review by benjie — CHANGES_REQUESTED** [review]
Excellent! The only issue is that the `src/assets/*` generated files have made it into the PR - please remove them!

(I will do squash and merge so no need to squash.)

**Comment by jmealo** [comment]
@benjie: I removed the built assets. How does this look?

**Review by benjie — APPROVED** [review]
Looks good to me! Thanks for your work on this 🙌

**Comment by benjie** [comment]
CI failed due to a network issue; retrying.

**Comment by jmealo** [comment]
@benjie looks good now. Maybe it was the AWS outage?

**Comment by benjie** [comment]
Thanks! 🙌

**Comment by benjie** [comment]
@jmealo Before we can officially support Postgres 15, PG15 will also need to be added to the graphile-engine repo. Now you know what's involved, are you interested in porting the changes back there?

**Comment by jmealo** [comment]
@Benjie Gillam ***@***.***> is their any react in that repo? That
was what made this a bit tricky IIRC.

**Comment by benjie** [comment]
I don't think so?

**Pull request merged**
